### PR TITLE
UCT/SM: add UCT_MD_FLAG_FIXED capability

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe @evgeny-leksikov 
sysv, posix and xpmem can all allocate memory with the user defined
base address.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4191/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4195/ for details (Mellanox internal link).

File: src/ucs/sys/sys.c [hunk 1]
```diff
@@ -536,6 +536,39 @@ ucs_status_t ucs_sysv_free(void *address)
     return UCS_OK;
 }
 
+ucs_status_t ucs_mmap_alloc(size_t *size, void **address_p,
```
Comment (yosefe) [inline src/ucs/sys/sys.c]: can we unite with ucs_sys_realloc

File: src/ucs/sys/sys.c [hunk 2]
```diff
@@ -536,6 +536,39 @@ ucs_status_t ucs_sysv_free(void *address)
...
+                            int flags UCS_MEMTRACK_ARG)
+{
+    size_t alloc_length;
+    void *addr;
+
+    alloc_length = ucs_align_up_pow2(*size, ucs_get_page_size());
+    if(*address_p != NULL) {
```
Comment (yosefe) [inline src/ucs/sys/sys.c]: code style: space after if

File: src/ucs/sys/sys.c [hunk 3]
```diff
@@ -536,6 +536,39 @@ ucs_status_t ucs_sysv_free(void *address)
...
+        flags |= MAP_FIXED;
+    }
+    addr = ucs_mmap(*address_p, alloc_length, PROT_READ | PROT_WRITE,
```
Comment (yosefe) [inline src/ucs/sys/sys.c]: pls add space line before

File: src/uct/base/uct_mem.c [hunk 4]
```diff
@@ -180,16 +177,18 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
...
         case UCT_ALLOC_METHOD_MMAP:
-            map_flags = uct_mem_get_mmap_flags(flags);
             /* Request memory from operating system using mmap() */
-            alloc_length = ucs_align_up_pow2(min_length, ucs_get_page_size());
-            address = ucs_mmap(addr, alloc_length, PROT_READ | PROT_WRITE,
-                               map_flags, -1, 0 UCS_MEMTRACK_VAL);
-            if (address != MAP_FAILED) {
+            map_flags = uct_mem_get_mmap_flags(flags);
+            alloc_length = min_length;
+            address = (flags & UCT_MD_MEM_FLAG_FIXED) ? addr : NULL;
```
Comment (yosefe) [inline src/uct/base/uct_mem.c]: need to drop this line to use addr as a hint as before

File: src/uct/sm/mm/mm_md.c [hunk 5]
```diff
@@ -34,6 +34,8 @@ ucs_status_t uct_mm_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
         return UCS_ERR_NO_MEMORY;
     }
 
+    seg->address = (flags & UCT_MD_MEM_FLAG_FIXED) ?  *address_p : NULL;
```
Comment (yosefe) [inline src/uct/sm/mm/mm_md.c]: IMHO it's better to pass 'flags' down to uct_mm_md_mapper_ops(md)->alloc() routine, than to use address!=NULL as indication

File: test/gtest/uct/test_mem.cc [hunk 6]
```diff
@@ -165,6 +166,42 @@ UCS_TEST_P(test_mem, mmap_fixed) {
...
+    uct_allocated_memory_t  uct_mem;
+    ucs_status_t            status;
+
+    n_success = 0;
+
+    for (i = 0; i < n_tryes; ++i) {
+        meth = (i % 2) ? UCT_ALLOC_METHOD_MMAP : UCT_ALLOC_METHOD_HUGE;
+
+        status = uct_mem_alloc((void *)p_addr, 1, UCT_MD_MEM_FLAG_FIXED,
+                &meth, 1, NULL, 0, "test", &uct_mem);
```
Comment (yosefe) [inline test/gtest/uct/test_mem.cc]: indentation

**Comment by yosefe** [comment]
@alinask pls review as well

**Comment by alinask** [comment]
:+1:

**Comment by yosefe** [comment]
tests are failing

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4204/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4218/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4220/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4221/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4227/ for details (Mellanox internal link).

**Pull request merged**